File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = set()
        n = len(nums)
        nums.sort()
        print(nums)
        for i in range(n):
            one = i + 1
            two = n - 1
            while one < two < n:
                val = nums[i] + nums[one] + nums[two]
                
                if val == 0:
                    res.add((nums[i], nums[one], nums[two]))
                    left = one + 1
                    while left < two < n and nums[one] == nums[left]:
                        # res.append((nums[i], nums[left], nums[two]))
                        # print('one', nums[i], nums[left], nums[two])
                        left += 1
                        
                    right = two - 1
                    while one < right < n and nums[two] == nums[right]:
                        # res.append((nums[i], nums[one], nums[right]))
                        # print('two', nums[i], nums[one], nums[right])
                        right -= 1
                    
                    one = left
                    two = right
                    
                elif val < 0:
                    one += 1
                else:
                    two -= 1
        return res

File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<Integer> neg = new ArrayList<>();
        List<Integer> pos = new ArrayList<>();
        int zero = 0;
        
        for (int num : nums) {
            if (num == 0) {
                zero++;
            } else if (num > 0) {
                pos.add(num);
            } else if (num < 0) {
                neg.add(num);
            }
        }
        
        Set<List<Integer>> res = new HashSet<>();
        Set<Integer> negSet = Set.copyOf(neg);
        Set<Integer> posSet = Set.copyOf(pos);
        if (zero > 0) {
            if (zero >= 3) {
                res.add(List.of(0,0,0));
            }

            for (int num : posSet) {
                if (negSet.contains(-num)) {
                    res.add(List.of(-num, 0, num));
                }
            }
        }
                
        for (int i = 0; i < neg.size(); i++) {
            for (int j = i + 1; j < neg.size(); j++) {
                if (posSet.contains(-neg.get(i)-neg.get(j))) {
                    res.add(List.of(Math.min(neg.get(i), neg.get(j)), Math.max(neg.get(i), neg.get(j)), -neg.get(i)-neg.get(j)));
                }
            }
        }
        
        for (int i = 0; i < pos.size(); i++) {
            for (int j = i + 1; j < pos.size(); j++) {
                if (negSet.contains(-pos.get(i)-pos.get(j))) {
                    res.add(List.of(-pos.get(i)-pos.get(j), Math.min(pos.get(i), pos.get(j)), Math.max(pos.get(i), pos.get(j))));
                }
            }
        }
        
        return List.copyOf(res);
    }
}